internal/rule/condition.go

```diff
@@ -3,82 +3,114 @@ package rule
 import (
 	"fmt"
 	"github.com/icinga/icinga-notifications/internal/event"
+	"github.com/icinga/icinga-notifications/internal/filter"
+	"math"
 	"time"
 )
 
+// RetryNever indicates that an escalation condition should never be retried once it has been evaluated.
+const RetryNever = time.Duration(math.MaxInt64)
+
 type EscalationFilter struct {
 	IncidentAge      time.Duration
 	IncidentSeverity event.Severity
 }
 
-func (c *EscalationFilter) EvalEqual(key string, value string) (bool, error) {
+// ReevaluateAfter returns the duration after which escalationCond should be reevaluated the
+// next time on the incident represented by e.
+//
+// escalationCond must correspond to an escalation that did not trigger on the incident
+// represented by e before. If nothing in the incident changes apart from time passing by,
+// the escalation is guaranteed to not trigger within the returned duration. After that
+// duration, the escalation should be reevaluated, and it may or may not trigger. If anything
+// else changes, for example due to an external event, the escalation must be reevaluated as
+// well.
+func (e *EscalationFilter) ReevaluateAfter(escalationCond filter.Filter) time.Duration {
+	retryAfter := RetryNever
+	for _, condition := range escalationCond.ExtractConditions() {
+		if condition.Column() == "incident_age" {
+			incidentAgeFilterValue, err := time.ParseDuration(condition.Value())
+			if err != nil {
+				continue
+			}
+
+			if incidentAgeFilterValue > e.IncidentAge {
+				retryAfter = min(retryAfter, incidentAgeFilterValue)
+			}
```

Comment:
D'oh, in my head, we already changed the return value to be relative to `e` (~now) instead of the incident start. Thus, I've written the comment with that in mind, but now it's actually inconsistent with the implementation.

Anyways, I'd change the implementation to match the comment, because I think this makes sense that way around. It can also be written a bit more compact without becoming harder to understand IMHO, maybe even a little improvement in that regard with no more `continue`.

```suggestion
			v, err := time.ParseDuration(condition.Value())
			if err == nil && v > e.IncidentAge {
				// The incident age is compared with a value in the future. Once that age is
				// reached, the escalation could trigger, so consider that time for reevaluation.
				retryAfter = min(retryAfter, v - e.IncidentAge)
			}
```

Of course, also requires a change where it's called to use the return value in the appropriate way.

Reply:
I somehow can't quite grasp why you keep bringing up how complex this is. It's only 2-3 lines of code that you don't need any comments to be able to understand it. But hey, that shouldn't matter, even if I didn't understand your comments from yesterday enough to understand what the function does compared to the previous ones. The comments just describe what might or might not happen when using this function somewhere in incident package.

Reply:
The main change request here is to do this, i.e. change the reference point for the return value to match the updated documentation:
```suggestion
			incidentAgeFilterValue, err := time.ParseDuration(condition.Value())
			if err != nil {
				continue
			}

			if incidentAgeFilterValue > e.IncidentAge {
				retryAfter = min(retryAfter, incidentAgeFilterValue - e.IncidentAge)
			}
```

But then I thought this can be simplified a bit and proposed a larger change.

Regarding the function documentation and inline comments: I think that those should give context to the code, they should say why and what for things are done. The function we're talking here has a very specific purpose: if I have an incident and an escalation that hasn't triggered yet, tell me for how long is it safe to ignore that escalation, because the result of evaluating the condition won't change until then. From the view of someone calling the function, that should be the relevant information, not what values are compared with the incident age in which way. I was trying to push you incrementally towards a direction where I think that is easier to understand from the interface and documentation of the function. I thought just requesting a big "just do it completely differently" change would have been frustrating for you, but my approach is already taking more iterations and time than I had expected, so it wasn't very successful either.